utils/src/main/java/software/amazon/awssdk/utils/async/AddingTrailingDataSubscriber.java

```diff
@@ -0,0 +1,148 @@
+ */
+@SdkProtectedApi
+public class AddingTrailingDataSubscriber<T> extends DelegatingSubscriber<T, T> {
+    private static final Logger log = Logger.loggerFor(AddingTrailingDataSubscriber.class);
+
+    /**
+     * The subscription to the upstream subscriber.
+     */
+    private Subscription upstreamSubscription;
+
+    /**
+     * The amount of unfulfilled demand the downstream subscriber has opened against us.
+     */
+    private final AtomicLong downstreamDemand = new AtomicLong(0);
+
+    /**
+     * Whether the upstream subscriber has called onComplete on us.
+     */
+    private volatile boolean onCompleteCalledByUpstream = false;
+
+    /**
+     * Whether the upstream subscriber has called onError on us.
+     */
+    private volatile boolean onErrorCalledByUpstream = false;
+
+    /**
+     * Whether we have called onComplete on the downstream subscriber.
+     */
+    private AtomicBoolean onCompleteCalledOnDownstream = new AtomicBoolean(false);
+
+    private final Supplier<T> trailingDataSupplier;
+    private volatile T trailingData;
+
+    public AddingTrailingDataSubscriber(Subscriber<? super T> subscriber,
+                                        Supplier<T> trailingDataSupplier) {
+        super(Validate.paramNotNull(subscriber, "subscriber"));
+        this.trailingDataSupplier = Validate.paramNotNull(trailingDataSupplier, "trailingDataSupplier");
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+
+        if (upstreamSubscription != null) {
+            log.warn(() -> "Received duplicate subscription, cancelling the duplicate.", new IllegalStateException());
+            subscription.cancel();
+            return;
+        }
+
+        upstreamSubscription = subscription;
+
+        subscriber.onSubscribe(new Subscription() {
+
+            @Override
+            public void request(long l) {
+                if (onErrorCalledByUpstream) {
+                    return;
+                }
+
+                if (onCompleteCalledByUpstream) {
+                    sendTrailingDataIfNeededAndComplete();
+                    return;
+                }
+
+                addDownstreamDemand(l);
+                upstreamSubscription.request(l);
+            }
+
+            @Override
+            public void cancel() {
+                upstreamSubscription.cancel();
+            }
+        });
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        onErrorCalledByUpstream = true;
+        subscriber.onError(throwable);
+    }
+
+    @Override
+    public void onNext(T t) {
+        Validate.paramNotNull(t, "item");
+        downstreamDemand.decrementAndGet();
+        subscriber.onNext(t);
+    }
+
+    @Override
+    public void onComplete() {
+        onCompleteCalledByUpstream = true;
+
+        trailingData = trailingDataSupplier.get();
+        if (trailingData == null || downstreamDemand.get() > 0) {
```

Comment:
We `get()` the trailing data from the supplier, but send it only if the supplier returns `null`? Is this really what we want?

Reply:
This code has been updated, but to answer your question, the original code allows it to trigger onComplete immediately if trailingData is null